graphql/handler/extension/introspection.go

```diff
@@ -27,3 +34,43 @@ func (c Introspection) MutateOperationContext(ctx context.Context, rc *graphql.O
 	rc.DisableIntrospection = false
 	return nil
 }
+
+func (c Introspection) InterceptField(ctx context.Context, next graphql.Resolver) (res interface{}, err error) {
+	res, err = next(ctx)
+
+	fc := graphql.GetFieldContext(ctx)
+	if fields, ok := res.([]introspection.Field); ok {
+		if c.AllowFieldFunc == nil {
+			return
+		}
+		t := fc.Parent.Result.(*introspection.Type)
+		var newFields []introspection.Field
+		for _, field := range fields {
+			allow, err := c.AllowFieldFunc(ctx, t, &field)
+			if err != nil {
+				return nil, err
+			}
+			if allow {
+				newFields = append(newFields, field)
+			}
+		}
+		res = newFields
+	} else if fields, ok := res.([]introspection.InputValue); ok {
+		if c.AllowInputValueFunc == nil {
+			return
+		}
+		t := fc.Parent.Result.(*introspection.Type)
```

Comment:
`intropection.Type`, `intropection.Directive` and `introspection.Field` will be coming.
https://spec.graphql.org/June2018/#sec-Schema-Introspection `__InputValue`

Reply:
I was thinking to change it to something like
```
type Introspection struct {
	Auth IntrospectionAuth
}

type IntrospectionAuth interface {
	AllowEnumValues(ctx context.Context, enumValues []introspection.EnumValue) ([]introspection.EnumValue, error)

	AllowFields(ctx context.Context, fields []introspection.Field) ([]introspection.Field, error)

	AllowInputValues(ctx context.Context, inputValues []introspection.InputValue) ([]introspection.InputValue, error)
}
```
and the allow func would be something like:
```
AllowFieldsFunc: func(ctx context.Context, fields []introspection.Field) ([]introspection.Field, error) {
						fc := graphql.GetFieldContext(ctx)
						newFields := make([]introspection.Field, 0, len(fields))
						if t, ok := fc.Parent.Result.(*introspection.Type); ok && *t.Name() == "TestType1" {
							for _, field := range fields {
								if field.Name == "TestField1" {
									continue
								}
								newFields = append(newFields, field)
							}
							return newFields, nil
						}
						return fields, nil
					},
```
This way the allow func is more verbose, but is called just one time and there is no need to have one function for each of intropection.Type, intropection.Directive and introspection.Field.
What do you think?